### ISSUE TITLE
mqtt AD dimmer status show level 99 instead of "on" in UI

### ISSUE THREAD

**Issue by TheUlpio** [domoticz/domoticz#5205]
Finally I switched from OZW to zwavejs2mqtt.
Obviously there are differences in how some controls react. One of them is that my dimmers, in UI, when are set full on, show the level 99 in the control instead of "on" as with OZW.
The behavior is correct in the off state (they show "off" instead of "0")

![image](https://user-images.githubusercontent.com/13163959/160372607-342b1fc3-4574-4c4d-9cb6-aff22082ba55.png)
This is the correct label when dimmer is off

![image](https://user-images.githubusercontent.com/13163959/160373058-3fba171f-304f-490c-be63-db4f5be9880b.png)
This is the imho wrong when the dimmer is full on, in this case I would expect an "on" label

**Comment by waltervl** [comment]
You failed to supply some needed info eg version, logs, device types.
You are showing the dashboard devices. Can you also show the same devices but now from the Switches Tab?

**Comment by rezzalex** [comment]
I confirm the behavior explained in First post.
But I don't see what is the impact / issue...

**Comment by TheUlpio** [comment]
@waltervl 
You're right, apologize me, this is my first issue here. Now I'm at work, this evening I'll post the needed data and screenshots.

@rezzalex 
Well, this morning I found my (little) daughter who insisted to turn on lights because they won't go to "on". Not a great issue indeed, mostly aesthetics, but imho:
1) it is not consistent with the other controls (e.g. blinds);
2) it appears differently than in the past;
3) ultimately it can be confusing.

**Comment by TheUlpio** [comment]
Here are the info I'm able to collect:
Domoticz: 2022.1 build 14285
zwavejs2mqtt: 6.5.2
zwave-js: 8.11.9

dimmer off:
![image](https://user-images.githubusercontent.com/13163959/160484170-fccab01a-3c78-4b0f-be2c-3e982d69c1e5.png)

dimmer on (wrong label):
![image](https://user-images.githubusercontent.com/13163959/160484348-7419a4b0-b6b8-4b33-8396-5c23abfdaa81.png)

Logs are not so explanatory:

>  2022-03-28 21:26:54.409 Status: User: Admin (IP: 192.168.1.163) initiated a switch command (1/Plafoniera corridoio/Off) 

>  2022-03-28 22:46:56.310 Status: User: Admin (IP: 192.168.1.163) initiated a switch command (1/Plafoniera corridoio/On)

**Comment by hestiap** [comment]
My 2 cts
it's the same since the beg. of zwavejs2mqtt w dz, not relative to a specific version
on Version: 2022.1 (build 14277)
zwavejs2mqtt log

> 2022-03-29` 08:21:28.119 INFO MQTT: Message received on zwave/150/38/0/targetValue/set: 'ON
2022-03-29 08:21:28.122 INFO ZWAVE: Writing 255 to 150-38-0-targetValue
2022-03-29 08:21:28.216 INFO ZWAVE: Node 150: value updated: 38-0-targetValue 0 => 255
2022-03-29 08:21:31.064 INFO ZWAVE: Node 150: value updated: 50-0-value-66049 0 => 4.7
2022-03-29 08:21:32.066 INFO ZWAVE: Node 150: value updated: 50-0-value-66049 4.7 => 12.1
2022-03-29 08:21:32.264 INFO ZWAVE: Node 150: value updated: 38-0-currentValue 0 => 99
2022-03-29 08:21:33.067 INFO ZWAVE: Node 150: value updated: 38-0-currentValue 99 => 99

![image](https://user-images.githubusercontent.com/73014953/160549597-d3d3c4c1-2367-4fb9-9b0b-e481153a76bc.png)

MQTT log
>zwave/150/38/0/targetValue
{"time":1648534888216,"value":255,"nodeName":"Allée","nodeLocation":"Jardin"}
zwave/150/38/0/currentValue
{"time":1648534893067,"value":99,"nodeName":"Allée","nodeLocation":"Jardin"}

I think the max is 99% for zwavejs2mqtt and on dz On is 100% or 255

**Comment by Rouzax** [comment]
Seeing the same thing here on Fibaro Dimmers, which is screwing up all sorts of dzvents scripts.
ZwaveJS2MQTT only goes from 0 to 99
![2022 05 20-11 12 54_ZWave_To_MQTT_—_Mozilla_Firefox_hI1KqWoC67](https://user-images.githubusercontent.com/4103090/169496086-005e9fe6-7113-443a-b153-4e342c143b38.gif)

zwavejs2mqtt: 6.9.1
zwave-js: 9.2.2
home id: 3244133030
home hex: 0xc15d8aa6

Domoticz: 2022.1 (build 14338)

**Comment by TheUlpio** [comment]
Could this be acceptable?
[dimmer.patch.txt](https://github.com/domoticz/domoticz/files/8847830/dimmer.patch.txt)

**Comment by Rouzax** [comment]
Is there any progress on this?

**Comment by waltervl** [comment]
Perhaps in https://github.com/domoticz/domoticz/issues/5217

**Comment by TheUlpio** [comment]
@waltervl no, looking at the code seems this issue isn't covered (not still tested anyway). The problem is, as @hestiap elaborate, the 99% issue (why 99? and not 100% for zwave2mqtt is a good question for that people). I tried to rebase the code on the 99% limit, but this broke zigbee devices, so no go. Maybe calculating limits based on max_level in mqtt payload and not hardcoding 100%, but imho this involves sql changes, so, for now I remain with my simple patch.

Another problem with the 99-100% difference is that the "int" in the code make the slider to jump an unit over the selected value if the selected level is over 50% (if I set a dimmer to 60%, it switches to 60%, then after a second, the slider jumps to 61%. This doesn't happen under 50% because the rounded value equals to the original number)

**Comment by Rouzax** [comment]
I suspect the reason for their (ZwaveJS2MQTT) choice is that the dimmers themselves go from 0 to 99
![image](https://user-images.githubusercontent.com/4103090/175524341-c9f47a84-e916-4ce9-8683-aa7a6f2856f7.png)

https://manuals.fibaro.com/content/manuals/en/FGD-212/FGD-212-EN-T-v1.3.pdf

With OpenZwave it had the same values (0-99), so there should already be a round up somewhere in the code that rounds 99% to 100% I suspect.

**Comment by Rouzax** [comment]
Not a programmer but I think this is it for the old OpenZwave implementation
https://github.com/domoticz/domoticz/blob/3179bc7f10aa55f2268f57168d51fed5f2ef12f6/hardware/ZWaveBase.cpp#L280

**Comment by CyrilB1531** [comment]
@TheUlpio I think this issue is fixed with The PR about Crazy Blinds. Is it possible for you to test it?

**Comment by TheUlpio** [comment]
@waltervl sorry for the late answer, but my little odroid takes ages to compile domoticz...
Well, yes and no. Yes because the status label becomes "on" instead of "99%" when the dimmer is on. Great work @CyrilB1531.
No because now the slider won't update its status in dashboard (only in dashboard. in switches tab it works as expected). I tried to clear the browse cache, run an anonymous instance of the browser and connect to domoticz with a new installed system. Always the same results. Anyway I'm investigating if this is only my issue, so this is another story.

Maybe this issue can be closed as it works for me.

**Comment by TheUlpio** [comment]
A side note.
If I set the slide of a dimmer device eg. from 70% to on (fully right wise) The lights goes to 100% status label shows 100% (not on), so, if I turn off the light and then turn it back on by clicking on the icon, light goes to 70% and not on.
I think this is due to the fact that 100% is interpreted as "last status", so, if I want an "on" lights, I have to set the slide to 99% (and this is impossible due to the "one unit jump" the slider performs, so always 98%).
This issue doesn't happen for blinds controls because there are dedicated buttons for close and open)

**Comment by CyrilB1531** [comment]
@TheUlpio thank you for your testing time, I will check that

**Comment by CyrilB1531** [comment]
I've made some changes yesterday about that, if you want to check it

**Comment by TheUlpio** [comment]
Thank you @CyrilB1531, with your last changes, now it seems that the blinds subsystem is totally decoupled from dimmer controls, so this issue is at it's original state (0% = off and 99% when on status)

**Comment by gizmocuz** [comment]
Could someone maybe test this pull request to see if this solves the issue?

https://github.com/domoticz/domoticz/pull/5300

**Comment by TheUlpio** [comment]
Thank you @gizmocuz I'll test your code this evening

**Comment by gizmocuz** [comment]
Thanks for testing... I am trying to replicate the issue but unfortunally i connected a switch (fibaro FGS) instead of a dimmer.... will try to find a dimmer tomorrow morning ;) ... they all look alike

**Comment by gizmocuz** [comment]
Is it possible to test with beta 14477 ... I think the problem was a internal rounding issue

**Comment by TheUlpio** [comment]
Sorry for the delay, but before starting the building of the patched Domoticz, I foolishly accepted a system upgrade and now the odroid I use for tests is bricked...
Just tried version 14777 on my production server and the control label now works as expected. Also set a level with the slider and then slide to "open" remember the open status end doesn't restore the previous level. Right!
this solve my previous comment:

> A side note. If I set the slide of a dimmer device eg. from 70% to on (fully right wise) The lights goes to 100% status label shows 100% (not on), so, if I turn off the light and then turn it back on by clicking on the icon, light goes to 70% and not on. I think this is due to the fact that 100% is interpreted as "last status", so, if I want an "on" lights, I have to set the slide to 99% (and this is impossible due to the "one unit jump" the slider performs, so always 98%). This issue doesn't happen for blinds controls because there are dedicated buttons for close and open)

so - at least for me - this version solves the issue, but I am still affected by this behavior:

> No because now the slider won't update its status in dashboard (only in dashboard. in switches tab it works as expected). I tried to clear the browse cache, run an anonymous instance of the browser and connect to domoticz with a new installed system. Always the same results. Anyway I'm investigating if this is only my issue, so this is another story.
 
 Maybe this issue can be closed?

**Comment by gizmocuz** [comment]
@TheUlpio , you do know that backups are made in the backup folder? Also when using docker you can checkout a specific version?

Just to recap this issue, all is working fine with lights/dimmer/slider now?

Because this line says yes and no:

"**Also set a level with the slider and then slide to "open" remember the open status end** doesn't restore the previous level. Right!"

_doesn't restore the previous level. Right! ?_

**Comment by TheUlpio** [comment]
@gizmocuz sorry for my bad English..

What I meant is that now dimmers work as expected:

- before (wrong behavior): if I set a level with the slider (e.g. 38%) and then push the button, lights turn off (from 38%), then move the slider to 99% (before your patch the maximum value), lights move according, of course, but if I turn off and on, lights don't turn to 99%, but to 38% (this was my "previous level" concept). No way to obtain (in Domoticz) the maximum level. Restoring the maximum level was possible only in the zwavejs2mqtt control panel.

- now (proper behavior) if I set a level, this is restored during power cycles (push the control button twice to turn off and on lights). If I drag the slider to 100% (the percentage is shown in the control label for a bit then changes to "on"), then during power cycles the "on" status is correctly applied.

The only thing left hanging is that the slider (only for dimmers, not e.g. for blinds controls and only in the dashboard, not in the switches tab where works as expected) is never updated: if I set the level to 38% then turn off lights, the slider remains to 38%.

**Comment by gizmocuz** [comment]
@TheUlpio , that is the expected behaviour, so you know when you turn it on, it will go to 38%, but I will check this there should be no difference... both options are ok, 0=off, and 38% when on... maybe that makes more sense... I will check how this worked on a very very old running installation and will make sure it stays the same

**Comment by gizmocuz** [comment]
@TheUlpio , found the issue, it was caused by a PR

https://github.com/domoticz/domoticz/commit/b71cad1cd3ee74e5f6f8119dc86a209c84c7b989

I made a quick patch and it should update correctly again in beta 14479 until the original author of the PR agrees or corrects it

Do note, the 'slider' was not updating on the dashboard, so if the value was 10% and on the switches tab it was set to 80%, on the dashboard it was not updating. This has been fixed
But turning the switch OFF still leaves the percentage the same (and does not go to zero), and I am quite sure that this was the default behavior since day 0, but i could be wrong...., this can be easily changed

**Comment by TheUlpio** [comment]
> I made a quick patch and it should update correctly again in beta 14479 until the original author of the PR agrees or corrects it

Thank you @gizmocuz for your time.
In the meanwhile I've restored my test system so I'm able to build from sources. Good.
Version 14479 (built from sources, I don't know if this version foresees your patch) doesn't work properly, but I've an update:

If light is off (slider fully leftwise) and I set a level, e.g.80%, the slider remains to 0% (not updating), but a page refresh, or returning from another tab shows the proper value. So maybe a refresh issue?

> Do note, the 'slider' was not updating on the dashboard, so if the value was 10% and on the switches tab it was set to 80%, on the dashboard it was not updating. This has been fixed But turning the switch OFF still leaves the percentage the same (and does not go to zero), and I am quite sure that this was the default behavior since day 0, but i could be wrong...., this can be easily changed

As long I remember (I use Domoticz since 2018) sliders always reflect the real and actual state of the device, not the value they would assume if I turn on the control. So, if the light is off, the slider is 0%; if the light is on, the slider is 100% and if light is set to level, the slider reflects that level.

**Comment by gizmocuz** [comment]
@TheUlpio , did you clear your browser cache ?
I just tested a dummy slider on a installation that runs 94 days... when turning off, the brightness slider stays on the same position

I emulated a dimmer via ESPHome, this node also transmits a empty color{} object that caused some trouble with the brightness level being set.

Because I also tweaked with blinds here, I don't want to make a new public beta yet. If you want I can create a custom build for you.
Please send me a message on Slack and specify for which platform (arm 32bit, 64bit, linux, windows)

**Comment by TheUlpio** [comment]
Apology me @gizmocuz. Yes, I've forgot the first rule... Clear the browser cache...

Your patch works flawlessly. Dimmer controls behavior is now perfect.

From my side this issue is resolved an can be closed

Thank you!

p.s. now i am able to compile again so if you need someone to run tests i am available

**Comment by gizmocuz** [comment]
@TheUlpio , great news! Is it possible to checkout this pull request to test the dimmers/blinds ?

https://github.com/domoticz/domoticz/pull/5318

**Comment by TheUlpio** [comment]
Sure!
Now I'm at work, but this evening I will launch the build with your patch

**Comment by gizmocuz** [comment]
@TheUlpio , could you try with the latest beta version, I just made a patch that could affect this behaviour

**Comment by TheUlpio** [comment]
I've took some time to be sure things works as expected. Domoticz version Beta 14496 built from sources.

Turn on, off and set levels from the wall switch work: device label and slider position react according with the commands. Good!

Turn on, of and set levels from the device itself work like a charm.

device is correctly updated also from json api calls like this:

> curl -k 'https://127.0.0.1/json.htm?type=command&param=switchlight&idx=14&switchcmd=Set%20Level&level=80'

Also commands from dzvents scripts update the device without problems.

as a last treat, the new rounding method resolve the +1 step issue I've described above, so, if the level 80% is set, the slider doesn't jump anymore to 81%, but remains to 80%. Very good!

@gizmocuz from my side this issue is fully resolved and can be closed. Feel free to ask further tests if you find them useful.
